server/base/src/main/java/org/apache/accumulo/server/fs/SpaceAwareVolumeChooser.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.accumulo.server.fs;
+
+import java.io.IOException;
+import java.util.NavigableMap;
+import java.util.Random;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.apache.accumulo.core.conf.AccumuloConfiguration;
+import org.apache.accumulo.core.conf.Property;
+import org.apache.accumulo.server.conf.ServerConfigurationFactory;
+import org.apache.commons.lang.StringUtils;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.FsStatus;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link VolumeChooser} that takes remaining HDFS space into account when making a volume choice
+ * rather than a simpler round robin.
+ */
+public class SpaceAwareVolumeChooser extends PreferredVolumeChooser {
+
+  public static final String HDFS_SPACE_RECOMPUTE_INTERVAL = Property.GENERAL_ARBITRARY_PROP_PREFIX
+      .getKey() + ".spaceaware.volume.chooser.recompute.interval";
+
+  // Time to wait in millis. Defaults to 5 min
+  private long computationCacheDuration = -1;
+  private long defaultComputationCacheDuration = 300000;
+  private long lastComputation = 0;
+  private static ConcurrentHashMap<String[],WeightedRandomCollection<String>> weightCache = new ConcurrentHashMap<>();
```

Comment:
I don't think using an array will work well for a map key.  The hashCode() and equals() functions for an array come from Object and therefore do not consider the array elements.   Could use `List<String>` as the key and construct the list using Arrays.asList(options) which efficiently wraps an array.  List created in this way have hashCode() and equals() based on the list elements.

Could use a Guava cache instead instead of using a ConcurrentHashMap .  Then the cache could be configured with an age.